hedera-node/src/main/java/com/hedera/services/txns/token/TokenDissociateTransitionLogic.java

```diff
@@ -55,50 +57,46 @@
 			TypedTokenStore tokenStore,
 			AccountStore accountStore,
-			TransactionContext txnCtx
+			TransactionContext txnCtx,
+			OptionValidator validator
 	) {
 		this.accountStore = accountStore;
 		this.tokenStore = tokenStore;
 		this.txnCtx = txnCtx;
+		this.validator = validator;
 	}
 
 	@Override
 	public void doStateTransition() {
 		/* --- Translate from gRPC types --- */
 		var op = txnCtx.accessor().getTxn().getTokenDissociate();
-		/* First the account */
-		final var grpcId = op.getAccount();
-		final var accountId = new Id(grpcId.getShardNum(), grpcId.getRealmNum(), grpcId.getAccountNum());
-		/* And then the tokens */
-		final List<Id> tokenIds = new ArrayList<>();
-		for (final var _grpcId : op.getTokensList()) {
-			tokenIds.add(new Id(_grpcId.getShardNum(), _grpcId.getRealmNum(), _grpcId.getTokenNum()));
-		}
+		final var accountId = Id.fromGrpcAccount(op.getAccount());
 
 		/* --- Load the model objects --- */
 		final var account = accountStore.loadAccount(accountId);
-		final List<TokenRelationship> tokenRelationships = new ArrayList<>();
-		for (final var tokenId : tokenIds) {
-			final var token = tokenStore.loadPossiblyDeletedToken(tokenId);
-			var tokenRelationShip = tokenStore.loadTokenRelationship(token, account);
-			tokenRelationships.add(tokenRelationShip);
+		final List<Pair<TokenRelationship, TokenRelationship>> tokenRelationships = new ArrayList<>();
+		for (final var tokenId : op.getTokensList()) {
+			final var token = tokenStore.loadPossiblyDeletedToken(Id.fromGrpcToken(tokenId));
+			var accountRelationship = tokenStore.loadTokenRelationship(token, account);
+			var treasuryRelationship = tokenStore.loadTokenRelationship(token, token.getTreasury());
+			tokenRelationships.add(Pair.of(accountRelationship, treasuryRelationship));
 		}
 
 		/* --- Do the business logic --- */
-		for(TokenRelationship tokenRelationship : tokenRelationships) {
-			var token = tokenRelationship.getToken();
-			var treasury = token.getTreasury();
-			var treasuryRelationShip = tokenStore.loadTokenRelationship(token, treasury);
-			tokenRelationship.validateAndDissociate(accountStore.getValidator(), treasuryRelationShip);
-			account.dissociateWith(List.of(token));
-
-			/* --- Persist the updated models --- */
-			accountStore.persistAccount(account);
-			tokenStore.persistTokenRelationships(List.of(treasuryRelationShip, tokenRelationship));
+		account.dissociateWith(tokenRelationships, validator);
+
+		/* --- Persist the updated models --- */
+		accountStore.persistAccount(account); // tx account
+		for (Pair<TokenRelationship, TokenRelationship> accountAndTreasuryPair : tokenRelationships) {
+			tokenStore.persistToken(accountAndTreasuryPair.getKey().getToken()); // token from tx token list
+			accountStore.persistAccount(accountAndTreasuryPair.getKey().getToken().getTreasury()); // treasury
+
+			tokenStore.persistTokenRelationships(List.of(accountAndTreasuryPair.getValue(), accountAndTreasuryPair.getKey()));
```

Comment:
Why don't we do only one `persistTokenRelationships` with all `TokenRelationship` entries making the List <Pair> flat and then do only one persist